File: src/components/island/metrics.ts
export const restingIsland = {
  width: 128,
  height: 32,
  radius: 16,
} as const;

export const peekIsland = {
  width: 220,
  height: 56,
  radius: 20,
} as const;

export const presentationIsland = {
  width: 332,
  height: 72,
  radius: 24,
} as const;

export const expandedIsland = {
  width: 360,
  height: 292,
  radius: 28,
} as const;

export const islandWindow = {
  width: 560,
  height: 520,
} as const;

export type Presence = "resting" | "peek" | "presentation" | "expanded";

export function metricsFor(presence: Presence) {
  switch (presence) {
    case "resting":
      return restingIsland;
    case "peek":
      return peekIsland;
    case "expanded":
      return expandedIsland;
    case "presentation":
      return presentationIsland;
  }
}

File: crates/bloop-core/src/metrics.rs
use serde::{Deserialize, Serialize};

pub const ABI_VERSION: &str = "1.0.0";
pub const ENGINE_VERSION: &str = "0.1.0";
pub const PACKAGE_FORMAT: u32 = 1;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Size {
    pub width: f64,
    pub height: f64,
    pub radius: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct IslandMetrics {
    pub resting: Size,
    pub peek: Size,
    pub presentation: Size,
    pub expanded: Size,
    pub window: Size,
}

impl IslandMetrics {
    pub fn load() -> Self {
        serde_json::from_str(include_str!("../../../shared/island-metrics.json"))
            .expect("island metrics")
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum Presence {
    Resting,
    Peek,
    Presentation,
    Expanded,
}

impl Presence {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Resting => "resting",
            Self::Peek => "peek",
            Self::Presentation => "presentation",
            Self::Expanded => "expanded",
        }
    }
}
